src/target/riscv/riscv-013.c

```diff
@@ -479,15 +470,25 @@ static int dtmcontrol_scan(struct target *target, uint32_t out, uint32_t *in_ptr
 	return ERROR_OK;
 }
 
-static void increase_dmi_busy_delay(struct target *target)
+static int increase_dmi_busy_delay(struct target *target)
 {
-	riscv013_info_t *info = get_info(target);
-	info->dmi_busy_delay += info->dmi_busy_delay / 10 + 1;
-	LOG_TARGET_DEBUG(target, "dtmcs_idle=%d, dmi_busy_delay=%d, ac_busy_delay=%d",
-			info->dtmcs_idle, info->dmi_busy_delay,
-			info->ac_busy_delay);
+	RISCV013_INFO(info);
+
+	int res = dtmcontrol_scan(target, DTM_DTMCS_DMIRESET,
+			NULL /* discard result */);
+	if (res != ERROR_OK)
+		return res;
 
-	dtmcontrol_scan(target, DTM_DTMCS_DMIRESET, NULL /* discard result */);
+	res = riscv_scan_increase_delay(&info->learned_delays,
+			RISCV_DELAY_BASE);
+	return res;
+}
+
+static void reset_learned_delays(struct target *target)
+{
+	RISCV013_INFO(info);
+	assert(info);
+	memset(&info->learned_delays, 0, sizeof(info->learned_delays));
```

Comment:
[Just a note; no change requested]

I typically prefer to reset all elements of a structure by assigning zeros to each member independently.

It is more typing, but the advantages are:

- static analysis tools are more likely to understand that than the memset
- when searching for all uses of variables (structure members), it is easy to miss the memset on the whole structure